### PR TITLE
Default set to turn off approximation_degree

### PR DESCRIPTION
**Pull request by nonhermitian**
### Summary
`approximation_degree` breaks unitary equivalence of the transpiler.  This sets the default to off unless explicitly set. 

### Details and comments

A related issue (which I have not modified here) is that both `approximation_degree=1.0` and `approximation_degree=0` turn off the feature due to a bug (per the documentation 1.0 turns it off).  However when thinking about "degree(s)" in other contexts, lower usually means less.  And therefore 1.0 should actually be maximum approximation rather than none.  This also aligns with boolean values.  I think the code should be changed to reflect this.

### CODE REVIEW

**Review by mtreinish — COMMENTED** [review]
I'm in favor of doing this because I agree this is a bad default. However, before we can merge this it needs a release note. Since this is an API change and users are potentially relying on the previous default (pretty much anyone running QV on hardware for real cares about this) the key thing for this PR is to have a release note that documents the change and provides a thorough explanation of why we're breaking things and how users can update if they desire the old default behavior. The process for adding release notes is here: https://github.com/Qiskit/qiskit-terra/blob/main/CONTRIBUTING.md#release-notes

**Comment by nonhermitian** [comment]
Yes.  However I did not want to write a note until the second point about definition of what degree means is resolved.  I am happy to make that change here as well.

**Comment by kdk** [comment]
> Yes. However I did not want to write a note until the second point about definition of what degree means is resolved. I am happy to make that change here as well.

I think saying something like "`approximation_degree` is being disabled" should be sufficient. (For what it's worth, `approximation_degree` is defined, albeit loosely, here

https://github.com/Qiskit/qiskit-terra/blob/367ed4e34a3d662ef8ed8f2b91e8823bba1b8ca8/qiskit/compiler/transpiler.py#L178

There's a bug in handling `approximation_degree` IIRC in the `TwoQubitBasisComposer` (by this point, `basis_fidelity`)

https://github.com/Qiskit/qiskit-terra/blob/dbc81a8dcdae9e9294b9058babf5883258daaa5d/qiskit/quantum_info/synthesis/two_qubit_decompose.py#L1087

that causes the transpiler to erroneously treat `approximation_degree=0.0` as `approximation_degree=None`).

**Comment by nonhermitian** [comment]
Is there one kind of release note preferred over others here?  I am thinking "other" but open to whatever type people want.

**Comment by jakelishman** [comment]
We use "upgrade" (the section heading is called "API changes" in the built notes) to describe things users should be aware of in the new version - that could be a choice?

**Comment by nonhermitian** [comment]
Ok, will change to "upgrade"

File: releasenotes/notes/turn-off-approx-degree-df3d39eb69f7f09f.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+---
+upgrade:
+  - |
+    This PR turns off the implicit use of `approximation_degree!=1.0` by default
```
Comment (mtreinish) [inline releasenotes/notes/turn-off-approx-degree-df3d39eb69f7f09f.yaml:4]:
```suggestion
    The implicit use of ``approximation_degree!=1.0`` by default in
```

File: releasenotes/notes/turn-off-approx-degree-df3d39eb69f7f09f.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+    in the the Qiskit transpiler at `optimization_level=3`.  The transpiler should, by default,
```
Comment (mtreinish) [inline releasenotes/notes/turn-off-approx-degree-df3d39eb69f7f09f.yaml:5]:
```suggestion
    in the :func:`~.transpile` function when ``optimization_level=3`` is set has been disabled.  The transpiler should, by default,
```

File: releasenotes/notes/turn-off-approx-degree-df3d39eb69f7f09f.yaml [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+    preserve unitarity of the input up to known transformations such as one-sided permutations
+    and similarity transformations.  This was broken by the previous use of `approximation_degree=None`
```
Comment (mtreinish) [inline releasenotes/notes/turn-off-approx-degree-df3d39eb69f7f09f.yaml:7]:
```suggestion
    and similarity transformations.  This was broken by the previous use of ``approximation_degree=None``
```

File: releasenotes/notes/turn-off-approx-degree-df3d39eb69f7f09f.yaml [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+    leading to incorrect results in cases such as Trotterized evolution with many time steps where
+    unitaries were being overly approximated leading to incorrect results.  It was decided that
+    transformations that break unitary equivilence should be explicitly activated by the user.
```
Comment (mtreinish) [inline releasenotes/notes/turn-off-approx-degree-df3d39eb69f7f09f.yaml:10]:
```suggestion
    transformations that break unitary equivalence should be explicitly activated by the user.
    If you desire the previous default behavior where synthesized :class:`~UnitaryGate` instructions
    are approximated up to the error rates of the target backend's native instructions you can explicitly
    set ``approximation_degree=None` when calling :func:`~.transpile` with ``optimization_level=3`, for
    example::
    
        transpile(circuit, backend, approximation_degree=None, optimization_level=3)
```

**Review by mtreinish — COMMENTED** [review]
This LGTM thanks for adding the release note. I just have some inline suggestions to make some small tweaks to the release note, then I think this should be good to go.

File: releasenotes/notes/turn-off-approx-degree-df3d39eb69f7f09f.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    The implicit use of ``approximation_degree!=1.0`` by default in
+    in the :func:`~.transpile` function when ``optimization_level=3`` is set has been disabled.  The transpiler should, by default,
+    preserve unitarity of the input up to known transformations such as one-sided permutations
+    and similarity transformations.  This was broken by the previous use of ``approximation_degree=None``
+    leading to incorrect results in cases such as Trotterized evolution with many time steps where
+    unitaries were being overly approximated leading to incorrect results.  It was decided that
+    transformations that break unitary equivalence should be explicitly activated by the user.
+    If you desire the previous default behavior where synthesized :class:`~UnitaryGate` instructions
+    are approximated up to the error rates of the target backend's native instructions you can explicitly
+    set ``approximation_degree=None` when calling :func:`~.transpile` with ``optimization_level=3`, for
```
Comment (mtreinish) [inline releasenotes/notes/turn-off-approx-degree-df3d39eb69f7f09f.yaml:13]:
```suggestion
    set ``approximation_degree=None` when calling :func:`~.transpile` with ``optimization_level=3``, for
```

**Review by mtreinish — APPROVED** [review]

**Pull request merged**